File: DSA/ShlokMadan_590017851/Day_6/Question1.java
public class Question1 {

    public static int rob(int[] nums) {
        int n = nums.length;

        if (n == 1) {
            return nums[0];
        }

        return Math.max(
                robLinear(nums, 0, n - 2),
                robLinear(nums, 1, n - 1)
        );
    }

    private static int robLinear(int[] nums, int start, int end) {
        int prev1 = 0;
        int prev2 = 0;

        for (int i = start; i <= end; i++) {
            int current = Math.max(prev1, prev2 + nums[i]);
            prev2 = prev1;
            prev1 = current;
        }

        return prev1;
    }

    public static void main(String[] args) {
        int[] nums = {1, 2, 3, 1};

        System.out.println(rob(nums));
    }
}

File: DSA/NabhanshRishiGaur_590028370/Day_6/Question_1.c
int robLinear(int* nums, int start, int end) {
    int prev1 = 0, prev2 = 0;
    for (int i = start; i <= end; i++) {
        int pick = prev2 + nums[i];
        int skip = prev1;
        int curr = (pick > skip) ? pick : skip;
        prev2 = prev1;
        prev1 = curr;
    }
    return prev1;
}

int rob(int* nums, int numsSize) {
    if (numsSize == 1) return nums[0];
    int case1 = robLinear(nums, 0, numsSize - 2);
    int case2 = robLinear(nums, 1, numsSize - 1);
    return (case1 > case2) ? case1 : case2;
}
